File: ISO-Tool/python/iso_tool/dependency_detector.py
"""Host dependency inventory shared by every ISO-Tool frontend.
This is deliberately stdlib-only so preflight works before optional packages exist.
"""
from __future__ import annotations
from dataclasses import asdict, dataclass
import json, os, shutil, subprocess, sys
from pathlib import Path

@dataclass
class Dependency:
    name: str
    commands: tuple[str, ...]
    required: bool = False
    category: str = "toolchain"
    path: str | None = None
    version: str | None = None
    status: str = "missing"
    detail: str = ""

DEFAULTS = [
    Dependency("git", ("git",), True, "source"),
    Dependency("python", ("python", "python3"), True, "runtime"),
    Dependency("cmake", ("cmake",), True, "build"),
    Dependency("gcc", ("gcc",), False, "compiler"), Dependency("g++", ("g++",), False, "compiler"),
    Dependency("msvc", ("cl",), False, "compiler"), Dependency("clang", ("clang",), False, "compiler"),
    Dependency("lld", ("lld",), False, "linker"), Dependency("nasm", ("nasm",), False, "assembler"),
    Dependency("masm", ("ml", "ml64"), False, "assembler"), Dependency("ninja", ("ninja",), False, "build"),
    Dependency("msbuild", ("msbuild",), False, "build"), Dependency("make", ("make",), False, "build"),
    Dependency("xorriso", ("xorriso", "xorrisofs"), False, "image"), Dependency("oscdimg", ("oscdimg",), False, "image"),
    Dependency("qemu", ("qemu-system-x86_64",), False, "verification"), Dependency("7zip", ("7z", "7zz"), False, "archive"),
    Dependency("zip", ("zip",), False, "archive"), Dependency("node", ("node",), False, "runtime"),
    Dependency("npm", ("npm",), False, "runtime"), Dependency("java", ("java",), False, "runtime"),
    Dependency("javac", ("javac",), False, "compiler"), Dependency("dotnet", ("dotnet",), False, "runtime"),
    Dependency("go", ("go",), False, "compiler"), Dependency("rustc", ("rustc",), False, "compiler"),
]

def _version(exe: str) -> str | None:
    for arg in ("--version", "-version"):
        try:
            p = subprocess.run([exe, arg], text=True, capture_output=True, timeout=5)
            text = (p.stdout or p.stderr).splitlines()
            if text: return text[0].strip()
        except (OSError, subprocess.SubprocessError): pass
    return None

def detect(extra: list[Dependency] | None = None) -> dict:
    deps = list(DEFAULTS) + list(extra or [])
    for d in deps:
        for command in d.commands:
            found = shutil.which(command)
            if found:
                d.path, d.version, d.status = found, _version(found), "found"
                break
        if d.status != "found" and d.required:
            d.detail = "Required dependency is not on PATH."
    return {"platform": sys.platform, "python": sys.version.split()[0], "path": os.environ.get("PATH", ""),
            "dependencies": [asdict(d) for d in deps],
            "summary": {"found": sum(d.status == "found" for d in deps), "missing": sum(d.status == "missing" for d in deps),
                        "required_missing": sum(d.status == "missing" and d.required for d in deps)}}

def write_report(output: Path) -> Path:
    output.parent.mkdir(parents=True, exist_ok=True); data = detect()
    output.write_text(json.dumps(data, indent=2), encoding="utf-8"); return output

if __name__ == "__main__":
    import argparse
    ap = argparse.ArgumentParser(); ap.add_argument("--output", default="dependency-report.json"); a = ap.parse_args()
    p = write_report(Path(a.output)); print(p)

File: ISO-Tool/vcpp/DependencyDetector.cpp
#include "DependencyDetector.hpp"
#include <sstream>
#include <array>
#ifdef _WIN32
#include <windows.h>
#endif
namespace iso_tool {
static std::string findCommand(const std::string& c) {
#ifdef _WIN32
    char buf[MAX_PATH]{}; DWORD n=SearchPathA(nullptr,c.c_str(),nullptr,MAX_PATH,buf,nullptr); if(n) return std::string(buf,n);
#else
    std::string q="command -v "+c+" 2>/dev/null"; std::array<char,512> b{}; FILE* f=popen(q.c_str(),"r"); if(!f)return {}; std::string s; if(fgets(b.data(),b.size(),f))s=b.data(); pclose(f); while(!s.empty()&&(s.back()=='\n'||s.back()=='\r'))s.pop_back(); return s;
#endif
    return {};
}
std::vector<Dependency> DetectDependencies(){
 const std::vector<Dependency> specs={
 {"git",{"git"},true,"source"},{"python",{"python","python3"},true,"runtime"},{"cmake",{"cmake"},true,"build"},
 {"gcc",{"gcc"},false,"compiler"},{"g++",{"g++"},false,"compiler"},{"msvc",{"cl"},false,"compiler"},{"clang",{"clang"},false,"compiler"},{"lld",{"lld"},false,"linker"},{"nasm",{"nasm"},false,"assembler"},{"masm",{"ml","ml64"},false,"assembler"},{"ninja",{"ninja"},false,"build"},{"msbuild",{"msbuild"},false,"build"},{"xorriso",{"xorriso","xorrisofs"},false,"image"},{"oscdimg",{"oscdimg"},false,"image"},{"qemu",{"qemu-system-x86_64"},false,"verification"},{"java",{"java"},false,"runtime"},{"javac",{"javac"},false,"compiler"},{"dotnet",{"dotnet"},false,"runtime"},{"node",{"node"},false,"runtime"},{"go",{"go"},false,"compiler"},{"rustc",{"rustc"},false,"compiler"},{"7zip",{"7z","7zz"},false,"archive"}};
 auto out=specs; for(auto& d:out){for(const auto& c:d.commands){d.path=findCommand(c);if(!d.path.empty()){d.status="found";break;}}if(d.status.empty())d.status="missing";}return out;
}
std::string DependenciesJson(const std::vector<Dependency>& ds){std::ostringstream o; o<<"{\"dependencies\":["; for(size_t i=0;i<ds.size();++i){if(i)o<<',';o<<"{\"name\":\""<<ds[i].name<<"\",\"required\":"<<(ds[i].required?"true":"false")<<",\"category\":\""<<ds[i].category<<"\",\"status\":\""<<ds[i].status<<"\",\"path\":\""<<ds[i].path<<"\"}";}o<<"]}";return o.str();}
}
